style: Sync changes from mozilla-central.

emilio opened the pull request:
See individual commits for details.

emilio committed:
style: Always restyle / repaint when a visited query finishes.

Differential Revision: https://phabricator.services.mozilla.com/D50810

emilio committed:
style: Fix ElementWrapper::is_link.

And do a full restyle only when the state goes from visited to unvisited or vice
versa. That is, use regular invalidation for addition or removals of href
attributes, for example.

Differential Revision: https://phabricator.services.mozilla.com/D50821

emilio committed:
style: Make zoom: 0 mean the same as zoom: 1.

This matches the WebKit implementation, and is clearly a violation of the rules
we generally use for ranges in CSS.

But it seems to be depended-on legacy behavior, see the linked WebKit bug, this
bug, and bug 1593317.

Differential Revision: https://phabricator.services.mozilla.com/D51539

emilio committed:
style: Simplify code for keeping alive shared memory until all sheets go away.

The existing code wasn't sound, as CSSOM objects also needed to go away before
the shared memory goes away (as they keep references to them).

This is sound assuming no presence of reference cycles introduced by CSSOM.

We may want to live with this and rely on chrome code not writing cycles like
this with UA stylesheet DOM objects.

We could explicitly drop all potentially-static objects... That seems pretty
error prone though.

Or we could also just leak the shared memory buffer, is there any reason why we
may not want to do that?

Differential Revision: https://phabricator.services.mozilla.com/D51870

nt1m committed:
style: Remove nsStackFrame platform code.

Differential Revision: https://phabricator.services.mozilla.com/D49487

emilio committed:
style: Use MaybeUninit in style struct clone impls / constructors.

Differential Revision: https://phabricator.services.mozilla.com/D51788

emilio committed:
style: Split collect_style_attribute_and_animation_rules.

Animations are their own cascade level per https://drafts.csswg.org/css-cascade-4/#cascade-origin

Differential Revision: https://phabricator.services.mozilla.com/D52575

emilio committed:
style: Optimize storage of ApplicableDeclaration again.

So that we don't regress perf.

Differential Revision: https://phabricator.services.mozilla.com/D52576

Commit:
style: Remove SpecialColorKeyword enum and merge into SystemColor.

Move all the entires of SpecialColorKeyword into SystemColor
and rearrange their computation to match.
Add the new SystemColor entries into the property list of nsXPLookAndFeel.

Differential Revision: https://phabricator.services.mozilla.com/D50903

emilio committed:
style: Fix cascade order of shadow parts.

This moves the shadow cascade order into the cascade level, and refactors the
code a bit for that.

Differential Revision: https://phabricator.services.mozilla.com/D49988

kvark committed:
style: Update some dependencies.

Differential Revision: https://phabricator.services.mozilla.com/D49458

hiikezoe committed:
style: Factor dynamic toolbar max height into layout metrics.

Now
* nsPresContext::mVisibleArea is excluding the toolbar max height so that
  ICB is now static regardless of the dynamic toolbar transition
* nsPresContext::mSizeForViewportUnits is introduced to resolve viewport units
  which is including the toolbar max height

That means that with the dynamic toolbar max height;

  mVisibleArea < mSizeForViewportUnits

See https://github.com/bokand/URLBarSizing for more detail backgrounds of this
change.

Depends on D50417

Differential Revision: https://phabricator.services.mozilla.com/D50418

emilio committed:
style: Remove NODE_MAY_BE_IN_BINDING_MNGR.

Never set anymore.

Differential Revision: https://phabricator.services.mozilla.com/D52994

emilio committed:
style: Centralize logic to ignore document colors.

This was a follow-up from the backplate stuff which I requested but didn't
happen.

Differential Revision: https://phabricator.services.mozilla.com/D53170

BorisChiou committed:
style: Treat 3d translate/scale as 2d if the value can be expressed as 2d.

For the individual transform properties if they spec a value that can be
expressed as 2d we treat as 2d and serialize accordingly.

We drop Translate::Translate and Scale::Scale, and then rename
Translate::Translate3D as Translate::Translate, Scale::Scale3D as
Scale::Scale. So now we use Translate::Translate to represent 2d and 3d
translation, and Scale::Scale to represent 2d and 3d scale. There is no
difference between 2d and 3d translate/scale in Gecko because we always
convert them into 3d format to layers (on the compositor thread), so this
change makes things simpler.

Differential Revision: https://phabricator.services.mozilla.com/D52931

BorisChiou committed:
style: Set WillChangeBits::TRANSFORM for offset-path and add tests for it.

Differential Revision: https://phabricator.services.mozilla.com/D53109

emilio committed:
style: Use only Origin during the cascade, rather than CascadeLevel.

The micro-benchmark `style-attr-1.html` regressed slightly with my patch, after
the CascadeLevel size increase.

This benchmark is meant to test for the "changing the style attribute doesn't
cause selector-matching" optimization (which, mind you, keeps working).

But in the process it creates 10k rules which form a perfect path in the rule
tree and that we put into a SmallVec during the cascade, and the benchmark
spends most of the time pushing to that SmallVec and iterating the declarations
(as there's only one property to apply).

So we could argue that the regression is minor and is not what the benchark is
supposed to be testing, but given I did the digging... :)

My patch made CascadeLevel bigger, which means that we create a somewhat bigger
vector in this case. Thankfully it also removed the dependency in the
CascadeLevel, so we can stop using that and use just Origin which is one byte to
revert the perf regression.

Differential Revision: https://phabricator.services.mozilla.com/D53181

emilio committed:
style: Remove NODE_IS_ANONYMOUS_ROOT.

We no longer have multiple kinds of anonymous subtrees, so we can get back one
node bit.

Differential Revision: https://phabricator.services.mozilla.com/D53344

BorisChiou committed:
style: Update the expected direction vector of rotate property in wpt.

For interpolations with the primitive rotate3d(), the direction vectors of
the transform functions get normalized first. This should also be applied to
rotate property.

https://drafts.csswg.org/css-transforms-2/#interpolation-of-transform-functions

Differential Revision: https://phabricator.services.mozilla.com/D52944

emilio committed:
style: Handle logical shorthand animations with variable references correctly.

When we physicalize the declarations for @keyframes, we end up having a physical
declaration with an unparsed value with `from_shorthand` being the logical
shorthand.

Account for this case properly when substituting custom properties, to avoid
panicking.

Differential Revision: https://phabricator.services.mozilla.com/D53663

emilio committed:
style: Minor debugging improvements.

These were useful when implementing forwarding, and forgot to send them
earlier.

Differential Revision: https://phabricator.services.mozilla.com/D53767

jeffin143 committed:
style: use Enum class for NS_STYLE_SHAPE_RENDERING instead of #define.

Differential Revision: https://phabricator.services.mozilla.com/D53840

emilio committed:
style: Make rust Atom use NonZeroUsize.

At first I thought this was going to enable simplifications in the selector
parser (to simplify the attribute selector setup), but I couldn't end up
shrinking the layout enough.

However this should help with bug 1559076, which returns Option<Atom>, and it
was easy to write.

Differential Revision: https://phabricator.services.mozilla.com/D53766

BorisChiou committed:
style: Add a preference for offset-path:ray().

Differential Revision: https://phabricator.services.mozilla.com/D53110

jeffin143 committed:
style: convert NS_STYLE_STROKE_LINECAP_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D53908

jeffin143 committed:
style: convert NS_STYLE_TEXT_ANCHOR_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D53956

emilio committed:
style: Invalidate parts in nested shadow trees correctly.

Differential Revision: https://phabricator.services.mozilla.com/D54010

emilio committed:
style: Implement shadow part forwarding (minus invalidation).

Some of the stuff, in particular inside GeckoBindings stuff should be
refactored to be less ugly and duplicate a bit less code, but the rest of the
code should be landable as is.

Some invalidation changes are already needed because we weren't matching with
the right shadow host during invalidation (which made existing ::part() tests
fail).

Pending invalidation work:

 * Making exportparts work right on the snapshots.
 * Invalidating parts from descendant hosts.

They're not very hard but I need to think how to best implement it:

 * Maybe get rid of ShadowRoot::mParts and just walk DOM descendants in the
   Shadow DOM.

 * Maybe implement a ElementHasExportPartsAttr much like HasPartAttr and use
   that to keep the list of elements.

 * Maybe invalidate :host and ::part() together in here[1]

 * Maybe something else.

Opinions?

[1]: https://searchfox.org/mozilla-central/rev/131338e5017bc0283d86fb73844407b9a2155c98/servo/components/style/invalidation/element/invalidator.rs#561

Differential Revision: https://phabricator.services.mozilla.com/D53730

Commit:
style: Don't accept two trailing <line-names> in the grid/grid-template shorthands.

Differential Revision: https://phabricator.services.mozilla.com/D53909

Commit:
style: Make some grid/grid-template values serialize to a shorter form.

Differential Revision: https://phabricator.services.mozilla.com/D53913

Commit:
style: Make 'grid-area'/'grid-column'/'grid-row' shorthands serialize to shortest possible form.

Differential Revision: https://phabricator.services.mozilla.com/D54386

jeffin143 committed:
style: convert NS_STYLE_OBJECT_FIT_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D54153

jeffin143 committed:
style: convert NS_STYLE_IME_MODE_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D54255

jeffin143 committed:
style: convert NS_STYLE_WINDOW_SHADOW_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D53920

emilio committed:
style: Simplify some code now that lifetimes are non-lexical.

Differential Revision: https://phabricator.services.mozilla.com/D54529

jeffin143 committed:
style: convert NS_STYLE_TEXT_ORIENTATION_* to an enum class in nsStyleConsts.h

Differential Revision: https://phabricator.services.mozilla.com/D54252

emilio committed:
style: Fix Servo_IsCssPropertyRecordedInUseCounter so that we also report disabled properties.

When zoom is disabled, we still count it, but with the current code the testing
function will throw instead of returning the right value, which means we'd fail
layout/style/test/test_use_counters.html.

Differential Revision: https://phabricator.services.mozilla.com/D55005

emilio committed:
style: Make Rust static atoms able to be used in const contexts.

I see atom dropping code generated in release builds for stuff like dropping the
"class" atom here:

https://searchfox.org/mozilla-central/rev/4df8821c1b824db5f40f381f48432f219d99ae36/servo/components/style/gecko/wrapper.rs#592

That is silly, and I hope making Atom be able to be used in const context will
help the compiler see that yeah, we're not doing anything interesting and the
atom shouldn't get dropped.

It also allows us to get rid of a few lazy_static!s, so we should do it anyway.

In order to accomplish this, compute the offset into gGkAtoms manually instead
of going through the static_atoms() array and then back to the byte offset.

Differential Revision: https://phabricator.services.mozilla.com/D55039

nordzilla committed:
style: Have scale function and scale property accept percentage value.

Differential Revision: https://phabricator.services.mozilla.com/D55012

emilio committed:
style: Don't specify minor version for cssparser.

emilio committed:
style: Undo minor debugging change.

emilio committed:
style: Rustfmt recent changes.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/Cargo.toml
 * @kichjang: components/script/Cargo.toml

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 9ff98da has been approved by `emilio`

<!-- @bors-servo r=emilio 9ff98da0b5dc0139e6a732272df7652626661a1a -->

bors-servo commented:
:hourglass: Testing commit 9ff98da0b5dc0139e6a732272df7652626661a1a with merge dd0fd3066c05a55cc35f7e604ec2f78b5f1e781d...

emilio committed:
Remove no-longer-needed arrayvec feature.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/BNvN2zuEQdi_ksI1UPiE4Q)

emilio committed:
style: Fix servo build.

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 7c3588c has been approved by `emilio`

<!-- @bors-servo r=emilio 7c3588cde189c85ceddfd10dc17c7b093193ea7a -->

bors-servo commented:
:hourglass: Testing commit 7c3588cde189c85ceddfd10dc17c7b093193ea7a with merge 14084fe60c3b064f0babd352f366a3c6a048a3a8...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/fDD-HoI5TUWRqtHIt8n9aQ)

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4053336 has been approved by `emilio`

<!-- @bors-servo r=emilio 4053336204511eea87d508753bd59cbdff244273 -->

bors-servo commented:
:hourglass: Testing commit 4053336204511eea87d508753bd59cbdff244273 with merge b05099e8b6fdf0197c97613056a856e11d822623...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZUkkP-NOQomS6Mz1pZWayQ)

emilio committed:
style: Update wpt expectations.

transform-scale-percent-001.html has been updated upstream, as the test is
invalid per CSSWG resolution.

emilio commented:
@bors-servo r+

 * Ugh, wpt update landed right before me.

bors-servo commented:
:pushpin: Commit 74e282e has been approved by `emilio`

<!-- @bors-servo r=emilio 74e282e5a99d01d840737317085008a415976867 -->

bors-servo commented:
:hourglass: Testing commit 74e282e5a99d01d840737317085008a415976867 with merge f7f0e148c6f1d49bf29c1e20d562f8400c48ad5a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ARmMueU1Sn2y-XUANejTig)

emilio commented:
@bors-servo r+

 * Missed a metadata file.

bors-servo commented:
:pushpin: Commit 6712e31 has been approved by `emilio`

<!-- @bors-servo r=emilio 6712e316dfdb0ef783c9f52bd336b027c187f879 -->

bors-servo commented:
:hourglass: Testing commit 6712e316dfdb0ef783c9f52bd336b027c187f879 with merge 1dd6ad7063a5bf71a37894654a7561338c6a39bc...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/M43xs7bLR6OUMlZAqWG9WQ)
Approved by: emilio
Pushing 1dd6ad7063a5bf71a37894654a7561338c6a39bc to master...
